### ISSUE TITLE
enhancement: fix true color check

### ISSUE THREAD

**Issue by stapelberg** [i3/i3#2452]
@Alexis211, please see @psychon’s comment in https://github.com/i3/i3/commit/287a0b4c3cc24a18f5063b7d5743f0b85d6852a1#commitcomment-18997285 — do we need to change the code?

**Comment by Alexis211** [comment]
@psychon's comment makes sense, and he probably knows more than me about XCB. I don't have time to look further into this issue right now, but I'll have a look at it next week if there is no progress.

**Comment by psychon** [comment]
TL;DR: https://github.com/awesomeWM/awesome/pull/2207/commits/1acae2aa6c7732f7224aa45c2dddd99c70963358#diff-9b11aa5e2c0ff4ee2cd21988e735676bR144

So, apparently people want transparent X11 window borders. I'd be surprised if there were no video driver where this runs into bugs, but at least it should work in theory. This made me recently write https://github.com/awesomeWM/awesome/pull/2207, which checks for a TrueColor / DirectColor visual and in that case handles color allocation without a round trip to the X11 server. For transparency, this handles depth=32 visuals specially and assumes that anything that is not red/green/blue is alpha.

I guess that something like this is needed here, too (by the way, I wonder what Xorg does on big endian machines and if this means that i3 messes up its color components there). I just wanted to keep the link here for reference on how I implemented something similar. Note that I do not feel like I really understands visuals either, so there might still be something wrong.

P.S.: Since I just wondered "what is the difference between direct color and true color"?
According to https://tronche.com/gui/x/xlib/window/visual-types.html, the color map for a direct color visual is writable. No idea what that is actually supposed to mean nor how it can work.

**Comment by Airblader** [comment]
Thanks for the information. I understand even less than you, but it's good to have it :-)
